### DISCUSSION TITLE
Querying the XDP program that is attached to an interface

### DISCUSSION THREAD

**Discussion by mpastyl** [cilium/ebpf#1380] [Q&A] [answered]
Hi, I would like to query which XDP program (if any) is attached to an interface and trying to figure out if there is support for that.

I managed to make it work by extending the wrappers around the main `BPF` syscall with a `LinkGetNextID`, similar to [ProgramGetNextID](https://github.com/cilium/ebpf/blob/42cbe8fe235c9e539efd59c940f0309813ce33e2/prog.go#L895) and I can open a PR if you are interested in that.
But I am wondering if I am missing a more obvious way to do it.

PS. There is a `QueryPrograms` that returns attached programs and any links they might have. Unfortunately, AFAIU, the [underlying syscall](https://elixir.bootlin.com/linux/v5.16.10/source/kernel/bpf/syscall.c#L3280) does not support the `BPF_XDP` attach type.

**Comment by florianl** [comment, marked as answer]
Without further information, it might be hard to get from a [Link](https://pkg.go.dev/github.com/cilium/ebpf/link#Link) to an interface. In addition, [Link](https://pkg.go.dev/github.com/cilium/ebpf/link#Link) is only supported for Linux kernel version 5.7 and newer iirc.

Instead of querying the `BPF` syscall, I would recomment to walk the interfaces using [`rtnetlink`](https://pkg.go.dev/github.com/jsimonetti/rtnetlink). As shown in the [example](https://pkg.go.dev/github.com/jsimonetti/rtnetlink#example-package-ListLink), one can query all interfaces quite easily and then further check the [`XDP`](https://pkg.go.dev/github.com/jsimonetti/rtnetlink#LinkXDP) element in the [`LinkAttributes`](https://pkg.go.dev/github.com/jsimonetti/rtnetlink#LinkAttributes) in the returned [`LinkMessage`](https://pkg.go.dev/github.com/jsimonetti/rtnetlink#LinkMessage). [`LinkXDP`](https://pkg.go.dev/github.com/jsimonetti/rtnetlink#LinkXDP) holds `FD`, which can be used as argument to [`NewProgramFromFD()`](https://pkg.go.dev/github.com/cilium/ebpf#NewProgramFromFD) to resolve the eBPF program.

**Reply by mpastyl** [reply]
Thanks a lot, `rtnetlink` looks really useful. 

> Without further information, it might be hard to get from a [Link](https://pkg.go.dev/github.com/cilium/ebpf/link#Link) to an interface

Since I am interested specifically for XDP links, shouldn't that work?
```Go
linkInfo, _ := somelink.Info()
if linkInfo.Type == link.TypeXDP {
	ifinxex := linkInfo.XDP().Ifindex
}
```

**Reply by lmb** [reply]
This works if you have a link from somewhere, for example if you pinned it into bpffs. It sounded like you don't have access to the link.

With a link, you don't have to check `.Type`:

```go
if xdp := linkInfo.XDP(); xdp != nil {
    // ...
}
```

**Reply by mpastyl** [reply]
Correct, I don't have easy access to the link. However, I was able to get it by extending the `link` package with a `LinkGetNextID()` and a `NewLinkFromID()`, which also includes extending the `sys` package a little bit to implement a bpf sycall for [BPF_LINK_GET_FD_BY_ID](https://github.com/cilium/ebpf/blob/65fb774468cbc51047f7115a57623832e8388eff/internal/sys/types.go#L112) and [BPF_LINK_GET_NEXT_ID](https://github.com/cilium/ebpf/blob/65fb774468cbc51047f7115a57623832e8388eff/internal/sys/types.go#L113).

So let me rephrase my original question, because I think I did a poor job explaining my intentions :smile: 
Are you interested in me opening a PR with the extensions mentioned above? Or should this use case better be handled with the help of another library, such as [rtnetlink](https://pkg.go.dev/github.com/jsimonetti/rtnetlink) as @florianl  pointed out.

**Reply by lmb** [reply]
Iterating bpf_links is definitely in scope, your contribution would be very welcome. I think it would make sense to base the code on https://github.com/cilium/ebpf/blob/65fb774468cbc51047f7115a57623832e8388eff/btf/handle.go#L200-L260 (but call it `link.Iterator` instead or something along those lines.)

**Reply by mpastyl** [reply]
Thanks a lot!

**Reply by ti-mo** [reply]
Something I didn't see mentioned in the thread: check out `link.QueryPrograms`. It will return a set of program IDs, either attached using netlink, prog_attach for cgroups, as well as everything attached using `bpf()` (tcx, xdp_link, netkit).

You won't be able to retrieve the link, but to check for XDP attachment to a specific interface in a generic way, I think it's your best bet. You can check out how we use it in cilium/cilium for inspiration.

Those link_get_next patches would still be welcome if you haven't sent them already!

**Reply by mpastyl** [reply]
> You can check out how we use it in cilium/cilium for inspiration.

Thanks for the hint, I will take a look.

> Those link_get_next patches would still be welcome if you haven't sent them already!

Sorry, I forgot to mention it here, but the changes have been merged :) https://github.com/cilium/ebpf/pull/1392
Let me close the discussion here, since it can now be done via iterating over the links.
